Show console completion options for tunings, extend and rename commands according to upstream

Robyt3 opened the pull request:
- Rename commands that list things according to upstream so they are consistent and improve command descriptions:
  - `dump_binds` -> `binds`
  - `tune_dump` -> `tunes`
  - `bans` (already named fittingly)
  - `dump_(local|remote)_console` (named correctly, `dump` should be used when writing to a file)
- Add success/error messages to `dump_(local|remote)_console` and improve description.
- Print error when using incorrect tuning names like `tune xyz 123`.
- Extend `tune_reset` so a single tuning value can be reset to default. For example `tune_reset player_collision` will reset just the `player_collision` tuning, whereas `tune_reset` will reset all to defaults like before.
- Get the actual tuning value after setting it, as the value that was applied may differ due to overflow or rounding errors.
- Extend `tune` so the current value of a given tuning variable is printed if no new value is given, so it works like commands in the console. For example `tune player_collision` will print the current value of the variable, whereas `tune player_collision 0` will change the value like before.
- Show completion options for tune params in console. Closes #5711.
  - Currently supports `tune`, `tune_reset` and `toggle_tune`.
  - For `tune_zone` the tune name is the second argument, so this is more difficult to handle.
  - Screenshot: 
![screenshot](https://user-images.githubusercontent.com/23437060/184015369-05853414-e45e-474d-a29a-90b82cf94159.png)

## Checklist

- [X] Tested the change ingame
- [X] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [X] Considered possible null pointers and out of bounds array indexing
- [X] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

def- commented:
Sounds good to me. Other opinions?

heinrich5991 commented:
Sounds good to me as well.

Do we have any maps that use `tune_reset`? Does any accidentally use it with a parameter so this change will break it?

Nit: In https://github.com/Robyt3/ddnet/commit/6cdc9e355088f3d84546955e07a0e3fdee58d836, you use `GetName` as a function name. I think we usually name getters without the "Get", so `Name` would probably be better?

Chairn commented on src/game/client/gameclient.cpp:
```diff
@@ -164,8 +164,8 @@ void CGameClient::OnConsoleInit()
 	Console()->Register("tune", "s[tuning] i[value]", CFGFLAG_SERVER, 0, 0, "Tune variable to value");
-	Console()->Register("tune_reset", "", CFGFLAG_SERVER, 0, 0, "Reset tuning");
-	Console()->Register("tune_dump", "", CFGFLAG_SERVER, 0, 0, "Dump tuning");
+	Console()->Register("tune_reset", "", CFGFLAG_SERVER, 0, 0, "Reset all tuning variables to defaults");
```
maybe should be named tunes_reset to keep it consistent with other commands that applies to all (binds|tunes|etc.)

Chairn replied:
Okay, nevermind, it's changed in following commit. Maybe split into 2 commands: tunes_reset for all, and tune_reset for single tuning?

Robyt3 replied:
I'd like to keep it simpler with a single command.

Chairn commented on src/game/client/components/binds.cpp:
```diff
@@ -254,7 +254,7 @@ void CBinds::OnConsoleInit()
 
 	Console()->Register("bind", "s[key] r[command]", CFGFLAG_CLIENT, ConBind, this, "Bind key to execute the command");
-	Console()->Register("dump_binds", "?s[key]", CFGFLAG_CLIENT, ConDumpBinds, this, "Print command executed by this keybindind or all binds");
+	Console()->Register("binds", "?s[key]", CFGFLAG_CLIENT, ConDumpBinds, this, "Print command executed by this keybinding or all binds");
```
Can you also rename ConDumpBinds accordingly to keep it consistent with tune_dump -> tunes?

Chairn commented on src/game/server/gamecontext.cpp:
```diff
@@ -3142,8 +3142,8 @@ void CGameContext::OnConsoleInit()
 	Console()->Register("toggle_tune", "s[tuning] i[value 1] i[value 2]", CFGFLAG_SERVER | CFGFLAG_GAME, ConToggleTuneParam, this, "Toggle tune variable");
-	Console()->Register("tune_reset", "", CFGFLAG_SERVER, ConTuneReset, this, "Reset tuning");
-	Console()->Register("tune_dump", "", CFGFLAG_SERVER, ConTuneDump, this, "Dump tuning");
+	Console()->Register("tune_reset", "", CFGFLAG_SERVER, ConTuneReset, this, "Reset all tuning variables to defaults");
```
Idem renaming to tunes

Chairn commented on src/game/client/components/console.cpp:
```diff
@@ -922,23 +928,28 @@ void CGameConsole::Toggle(int Type)
 void CGameConsole::Dump(int Type)
 {
 	CInstance *pConsole = Type == CONSOLETYPE_REMOTE ? &m_RemoteConsole : &m_LocalConsole;
+	char aBuf[256];
```
I would add IO_MAX_PATH_LENGTH to 256 to prevent cases where storage.cfg points to a long path and aBuf message being truncated

Chairn commented on src/game/client/components/console.cpp:
```diff
@@ -171,6 +180,26 @@ void CGameConsole::CInstance::PossibleCommandsCompleteCallback(int Index, const
+		char aBuf[512];
+		const char *pSrc = pInstance->GetString();
+		size_t i = 0;
+		for(; i < sizeof(aBuf) - 2 && pSrc[i] && pSrc[i] != ' '; i++)
```
Why not using str_find then str_copy?

Chairn commented on src/game/client/components/console.cpp:
```diff
@@ -386,17 +439,26 @@ void CGameConsole::CInstance::OnInput(IInput::CEvent Event)
-				aBuf[i] = *pSrc;
-			aBuf[i] = 0;
+			size_t i = 0;
+			for(; i < sizeof(aBuf) - 1 && pSrc[i] && pSrc[i] != ' '; i++)
```
same here, str_find + str_copy?

Chairn commented:
I tested the PR, it works fine, thanks for the great work. Note that when changing a tune, message is printed twice.

def- commented:
> Do we have any maps that use `tune_reset`? Does any accidentally use it with a parameter so this change will break it?

tune_reset is not a valid map setting

Robyt3 commented:
> Note that when changing a tune, message is printed twice.

Fixed.

> I think we usually name getters without the "Get", so Name would probably be better?

Fixed.

I also applied the changes from the commits `Print incorrect tuning parameter name on error` and `Print the actual tuning value after setting it` to the other ddnet specific commands as well.

Chairn reviewed (approved):
:+1:

def- commented:
bors r+
Thanks @Robyt3 for the change, thanks @Chairn for reviewing.
